src/utils/view.test.js

```diff
@@ -0,0 +1,407 @@
+
+      const extents = {
+        translate: {
+          minX: -110,
+          minY: -120,
+          maxX: 120,
+          maxY: 130
+        },
+        scale: {
+          minK: 1,
+          maxK: 1.5
+        }
+      };
+
+      setViewExtents(view, extents);
+
+      // Attempt going beyond max extents and max scale
+      setViewTransform(
+        view,
+        { x: 500, y: 500, k: extents.scale.maxK * 2 },
+        0,
+        false
+      );
+
+      // Result should have constrained to max extents * max scale
+      const v = getViewTransform(view);
+      expect(v.x).toEqual(extents.translate.maxX * extents.scale.maxK);
+      expect(v.y).toEqual(extents.translate.maxY * extents.scale.maxK);
+      expect(v.k).toEqual(extents.scale.maxK);
+
+      // Attempt going beyond min extents and min scale
+      setViewTransform(
+        view,
+        { x: -500, y: -500, k: extents.scale.minK * 0.5 },
+        0,
+        false
+      );
+
+      // Result should have constrained to min extents * min scale
+      const v2 = getViewTransform(view);
+      expect(v2.x).toEqual(extents.translate.minX * extents.scale.minK);
+      expect(v2.y).toEqual(extents.translate.minY * extents.scale.minK);
+      expect(v2.k).toEqual(extents.scale.minK);
+    });
+  });
+
+  describe('getViewExtents, setViewExtents', () => {
+    it('getViewExtents returns initial infinite extents', () => {
+      const view = mockView(100, 100);
+
+      expect(getViewExtents(view)).toEqual({
+        translate: {
+          minX: -Infinity,
+          maxX: Infinity,
+          minY: -Infinity,
+          maxY: Infinity
+        },
+        scale: {
+          minK: 0,
+          maxK: Infinity
+        }
+      });
+    });
+
+    it('setViewExtents applies given extents and getViewExtents returns same', () => {
+      const view = mockView(100, 100);
+
+      const extents = {
+        translate: {
+          minX: -1,
+          maxX: 2,
+          minY: -3,
+          maxY: 4
+        },
+        scale: {
+          minK: 0.5,
+          maxK: 0.6
+        }
+      };
+
+      setViewExtents(view, extents);
+      expect(getViewExtents(view)).toEqual(extents);
+    });
+  });
+
+  describe('viewTransformToFit', () => {
+    it('returns expected transform when example must be scaled down to fit', () => {
+      // Example where object is too large, so scale down
+      const minScaleFocus = 0.8;
+      const transform = viewTransformToFit({
+        offset: { x: 10, y: 10 },
+        focus: { x: 50, y: 50 },
+        viewWidth: 200,
+        viewHeight: 100,
+        objectWidth: 150,
+        objectHeight: 300,
+        minScaleX: 0.4,
+        minScaleFocus,
+        focusOffset: 0.8
+      });
+
+      expect(transform).toEqual({
+        k: minScaleFocus,
+        x: -50,
+        y: 60
+      });
+    });
+
+    it('returns expected transform when example must be scaled up to fit', () => {
+      // Example where object is too small, so scale up
```

Comment:
This might need a bit more clarification to clarify what ratio is viewed as a small object for scaling up